### PR TITLE
Add GDX recalculation for TokenPurchased

### PR DESCRIPTION
**Pull request by Shloyem**
Re-using the old GDX airdrop script for the new reserve. 
Plz ignore some naming/logging/unused variables at the moment

### CODE REVIEW

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 1]
```diff
@@ -198,3 +198,180 @@ export const airdrop = (ethers: typeof Ethers, ethSnapshotBlock) => {
...
 };
+
+export const airdropNew = (ethers: typeof Ethers, ethSnapshotBlock) => {
+  const getBuyingAddresses = async (addresses = {}, isContracts = {}) => {
+    const provider = new ethers.providers.InfuraProvider();
+
+    let newReserve = await ethers.getContractAt(
+      [
+        "event TokenPurchased(address indexed caller,address indexed inputToken,uint256 inputAmount,uint256 actualReturn,address indexed receiverAddress)",
+        "event TokenSold(address indexed caller,address indexed outputToken,uint256 gdAmount,uint256 contributionAmount,uint256 actualReturn,address indexed receiverAddress)"
```
Comment (Shloyem) [inline scripts/gdx/gdxAirdropCalculation.ts:209]: TokenSold to be removed and replaced with an exchangehelper call, similar to the first airdrop that's above

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 2]
```diff
@@ -198,3 +198,180 @@ export const airdrop = (ethers: typeof Ethers, ethSnapshotBlock) => {
...
+        // Print out all the values:
+        const ps = logs.map(async log => {
+          let isContract =
+            (await newReserve.provider
+              .getCode(log.args.caller)
+              .catch(e => "0x")) !== "0x";
+          let balance = addresses[log.args.caller] || 0;
+          console.log({balance});
+          console.log(`actualReturn: ${log.args.actualReturn.toNumber()}`);
+          addresses[log.args.caller] =
```
Comment (sirpy) [inline scripts/gdx/gdxAirdropCalculation.ts:269]: this should be token purchased

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 3]
```diff
@@ -198,3 +199,189 @@ export const airdrop = (ethers: typeof Ethers, ethSnapshotBlock) => {
...
+      fs.readFileSync("airdrop/buyBalancesCombined.json").toString()
+    );
+
+    let toTree: Array<[string, number]> = Object.entries(addressesCombined).map(
+      ([addr, gdx]) => {
+        return [addr, gdx as number];
+      }
+    );
+
+    toTree = sortBy(toTree, "1").reverse();
```
Comment (Shloyem) [inline scripts/gdx/gdxAirdropCalculation.ts:318]: Need to add BigNumber sorting

File: hardhat.config.ts [hunk 4]
```diff
@@ -197,6 +197,22 @@ task("gdxAirdrop", "Calculates airdrop data")
     }
   });
 
+task("gdxAirdropNew", "Calculates new airdrop data")
```
Comment (sirpy) [inline hardhat.config.ts:200]:
```suggestion
task("gdxAirdropRecover", "Calculates new airdrop data")
```

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 5]
```diff
@@ -198,3 +210,224 @@ export const airdrop = (ethers: typeof Ethers, ethSnapshotBlock) => {
...
+    fs.writeFileSync(
+      "airdrop/gdxairdropNew.json",
+      JSON.stringify({ treeData, merkleRoot })
+    );
+  };
+
+  const addCalculationsToPreviousData = async () => {
+    // get previous airdrop and turn it into BigNumbers
+    const { addresses, isContracts } = JSON.parse(
+      fs.readFileSync("airdrop/buyBalances.json").toString()
```
Comment (sirpy) [inline scripts/gdx/gdxAirdropCalculation.ts:391]: is that the file the issue specify to use?

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 6]
```diff
@@ -198,3 +210,224 @@ export const airdrop = (ethers: typeof Ethers, ethSnapshotBlock) => {
...
+        return [addr, gdx as BigNumber];
+      }
+    );
+
+    // Convert minuses to 0
+    newAddressesArray.forEach((a,_) => { a[1] = (a[1] < ZERO) ? ZERO : a[1]});
+
+    // console.log(`Before sorting`);
+    // newAddressesArray.forEach((a,_) => { console.log(`${a[0].toString()}:${ethers.BigNumber.from(a[1]).toString()}\n`)}); 
+    newAddressesArray.sort((a, b) => BigNumber.from(a[1]).sub(b[1]) < ZERO ? 0 : -1 );
```
Comment (sirpy) [inline scripts/gdx/gdxAirdropCalculation.ts:412]: bignumber.gt() ?

Comment (sirpy) [reply scripts/gdx/gdxAirdropCalculation.ts:412]: also this sort doesnt do anything... result is not saved

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 7]
```diff
@@ -198,3 +210,224 @@ export const airdrop = (ethers: typeof Ethers, ethSnapshotBlock) => {
...
+    const { newAddresses, newIsContracts }  = await getHoldersInformation();
+
+    let newAddressesArray: Array<[string, BigNumber]> = Object.entries(newAddresses).map(
+      ([addr, gdx]) => {
+        return [addr, gdx as BigNumber];
+      }
+    );
+
+    // Convert minuses to 0
+    newAddressesArray.forEach((a,_) => { a[1] = (a[1] < ZERO) ? ZERO : a[1]});
```
Comment (sirpy) [inline scripts/gdx/gdxAirdropCalculation.ts:408]: use filter

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 8]
```diff
@@ -198,3 +210,224 @@ export const airdrop = (ethers: typeof Ethers, ethSnapshotBlock) => {
...
+    // newAddressesArray.forEach((a,_) => { console.log(`${a[0].toString()}:${ethers.BigNumber.from(a[1]).toString()}\n`)});
+
+    // Unite previous airdrop with current information 
+    for (const newAddressEntry of newAddressesArray) {
+      const address = newAddressEntry[0];
+      const addition = newAddressEntry[1];
+      // console.log({address})
+      // console.log({before: BigNumber.from(addressesCombined[address] || "0").toString() })
+      // console.log({addition: addition.toString()})
+      addressesCombined[address] = ethers.BigNumber.from(addressesCombined[address] || 0).add(addition.toString());
```
Comment (sirpy) [inline scripts/gdx/gdxAirdropCalculation.ts:423]:
```suggestion
      addressesCombined[address] = ethers.BigNumber.from(addressesCombined[address] || 0).add(addition.toString()).toString();
```

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 9]
```diff
@@ -198,3 +210,224 @@ export const airdrop = (ethers: typeof Ethers, ethSnapshotBlock) => {
...
+    newAddressesArray.forEach((a,_) => { a[1] = (a[1] < ZERO) ? ZERO : a[1]});
+
+    // console.log(`Before sorting`);
+    // newAddressesArray.forEach((a,_) => { console.log(`${a[0].toString()}:${ethers.BigNumber.from(a[1]).toString()}\n`)}); 
+    newAddressesArray.sort((a, b) => BigNumber.from(a[1]).sub(b[1]) < ZERO ? 0 : -1 );
+      // console.log(`After sorting`);
+    // newAddressesArray.forEach((a,_) => { console.log(`${a[0].toString()}:${ethers.BigNumber.from(a[1]).toString()}\n`)});
+
+    // Unite previous airdrop with current information 
+    for (const newAddressEntry of newAddressesArray) {
```
Comment (sirpy) [inline scripts/gdx/gdxAirdropCalculation.ts:417]: i think you can just iterate over newAddresses here, and skip here gdx balances that are <=0

**Review by sirpy — CHANGES_REQUESTED** [review]

File: hardhat.config.ts (hunk 4 above)
Comment (Shloyem) [reply hardhat.config.ts:200]: Done + references

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 10]
```diff
@@ -198,3 +210,224 @@ export const airdrop = (ethers: typeof Ethers, ethSnapshotBlock) => {
...
+          await Promise.all([...purchasedEventsMapped, ...soldEventsMapped]);
+        });
+        await Promise.all(processedChunks);
+      }
+    };
+
+    await populateListOfAddressesAndBalances(newReserve, reserveTokenPurchasedFilter, reserveTokenSoldFilter);
+    await populateListOfAddressesAndBalances(exchangeHelper, exchangeHelperTokenPurchasedFilter, exchangeHelperTokenSoldFilter);
+
+    delete newAddresses[exchangeHelper.address];
```
[the remark this answers is not in the document]
Comment (Shloyem) [reply scripts/gdx/gdxAirdropCalculation.ts:312]: true, there aren't, removing the line

File: scripts/gdx/gdxAirdropCalculation.ts (hunk 5 above)
Comment (Shloyem) [reply scripts/gdx/gdxAirdropCalculation.ts:391]: gdxairdrop.json is, changing to it

File: scripts/gdx/gdxAirdropCalculation.ts (hunk 6 above)
Comment (Shloyem) [reply scripts/gdx/gdxAirdropCalculation.ts:412]: gt() is better, taking it for another sorting, this one is removed was just for printing the highest diff

File: scripts/gdx/gdxAirdropCalculation.ts (hunk 9 above)
Comment (Shloyem) [reply scripts/gdx/gdxAirdropCalculation.ts:417]: Good idea

File: scripts/gdx/gdxAirdropCalculation.ts (hunk 8 above)
Comment (Shloyem) [reply scripts/gdx/gdxAirdropCalculation.ts:423]: done

File: scripts/gdx/gdxAirdropCalculation.ts [hunk 11]
```diff
@@ -329,17 +327,14 @@ export const airdropNew = (ethers: typeof Ethers) => {
     // console.log(`Before sorting`);
     // toTree.forEach((a,_) => { console.log(`${a[0].toString()}:${ethers.BigNumber.from(a[1]).toString()}\n`)}); 
     
-    toTree.sort((a, b) => BigNumber.from(a[1]).sub(b[1]) < ZERO ? 0 : -1 );
+    toTree.sort((a, b) => BigNumber.from(b[1]).gt(a[1]) ? 0 : -1 );
```
Comment (sirpy) [inline scripts/gdx/gdxAirdropCalculation.ts:330]: this line still does nothing

**Review by sirpy — APPROVED** [review]

**Pull request merged**